src/m3ninx/index/segment/builder/builder.go

```diff
@@ -247,3 +302,11 @@ func (b *builder) Terms(field []byte) (segment.TermsIterator, error) {
 
 	return newTermsIter(terms.uniqueTerms), nil
 }
+
+func (b *builder) Close() error {
+	b.closed.Store(true)
+	for _, q := range b.indexQueues {
+		close(q)
```

Comment:
This is racy since `b.closed.Load()` might happen before you set the `closed.Store(true)` on line 307 and it thinks it's not closed but then you close and close each queue.  

I would instead perhaps just create a `builderStatus` struct with a `sync.RWMutex` and hold a read lock the whole time you are in `InsertBatch`. Acquiring an un-contended read lock is usually only slightly more expensive (still just low tens of nanoseconds) as an atomic store/load:
```go

type builder struct {
  status builderStatus
}

type builderStatus struct {
  sync.RWMutex 
  closed bool
}

func (b *builder) InsertBatch(...) error {
  b.status.RLock()
  defer b.status.RLock()

  if b.status.closed {
    return errClosed
  }

  // now protected the whole time enqueing for the whole batch
}

func (b *builder) Close() error {
  b.status.Lock()
  defer b.status.Unlock()
  
  b.status.closed = true
  for _, q := range b.indexQueues {
    close(q)
  }
}

```